src/Symfony/Component/Cache/Adapter/TagAwareAdapter.php

```diff
@@ -363,39 +364,39 @@ private function getTagVersions(array $tagsByKey, bool $persistTags): array
         $tags = [];
         foreach ($tagVersions as $tag => $version) {
             $tags[$tag.static::TAGS_PREFIX] = $tag;
+            if ($fetchTagVersions) {
+                continue;
+            }
             $knownTagVersion = $this->knownTagVersions[$tag] ?? [0, null];
-            if ($fetchTagVersions || $knownTagVersion[1] !== $version || $now - $knownTagVersion[0] >= $this->knownTagVersionsTtl) {
+            if ($knownTagVersion[1] !== $version || $now >= $knownTagVersion[0]) {
                 // reuse previously fetched tag versions up to the ttl
                 $fetchTagVersions = true;
             }
-            unset($this->knownTagVersions[$tag]); // For LRU tracking
-            if ([0, null] !== $knownTagVersion) {
-                $this->knownTagVersions[$tag] = $knownTagVersion;
-            }
         }
 
         if (!$fetchTagVersions) {
             return $tagVersions;
         }
 
+        if (self::MAX_NUMBER_OF_KNOWN_TAG_VERSIONS < \count($this->knownTagVersions)) {
```

Comment:
We could postpone pruning for a few cycles to give all tags in cache a chance to be reused.

```suggestion
        static $counter = 0;
        if (self::MAX_NUMBER_OF_KNOWN_TAG_VERSIONS < \count($this->knownTagVersions) && !++$counter % 10) {
```

Reply:
I found better solution - to use FIFO instead of LRU, and to cut it off only when the first tag is expired.

Reply:
I like it thanks